pkg/resource/deploy/step_generator.go

```diff
@@ -965,6 +965,12 @@ func (sg *stepGenerator) providerChanged(urn resource.URN, old, new *resource.St
 		return false, err
 	}
 
+	if alias, ok := sg.aliased[oldRef.URN()]; ok && alias == newRef.URN() {
```

Comment:
Are there any alias cases we need to consider beyond `alias == newRef.URN()`?

Reply:
I must admit I'm new to this code.

I tested it with an alias `Type = "pulumi:providers:azure-nextgen"` and it hit the condition with

> an aliased provider from "urn:pulumi:dev::azure-nextgen-cs::pulumi:providers:azure-nextgen::p1" to "urn:pulumi:dev::azure-nextgen-cs::pulumi:providers:azure-native::p1"

so I assumed that aliases are resolved prior to this point.

Reply:
Aliases should be fully resolved by this point. The only thing I'm not entirely sure of is whether or not we should also check that the ID of the provider has not changed, since that would imply that the provider is different and should trigger replacement.

Reply:
@pgavlin I tested renaming the provider object (`p1` to `p2` in the example above) and that did trigger the replacement. Or do you mean something different?

Reply:
I can't seem to alias the provider with a `urn`: `Urn = "urn:pulumi:dev::azure-nextgen-cs::pulumi:providers:azure-native::p1"}` causes

> error: unknown provider 'urn:pulumi:dev::azure-nextgen-cs::pulumi:providers:azure-native::p2::dd7b6725-6649-4ce5-9429-2ff9053b920b' for resource 'urn:pulumi:dev::azure-nextgen-cs::azure-native:resources/latest:ResourceGroup::test-delete-rg'

But this doesn't look related (getting the same with 1.21.2)